### PR TITLE
Add CSV Subscriptions Import

### LINKED ISSUE #2319: [Bug] Unexpected char 'C' at line 1 - Youtube is now serving csv file instead of json file when exporting the subscriptions
TEMPORARY SOLUTION: https://github.com/iv-org/invidious/issues/2319#issuecomment-900358678

-------

YouTube gives a csv file instead of a json file when exporting the subscriptions.

------

**Describe the bug**
I sadly can't import my YT Subs to Invie

**Steps to Reproduce**
Import your YT Subs to Invidious while having Google on German

**Logs**
Title: `Unexpected char 'K' at line 1, column 1 (JSON::ParseException)`
Date: `2021-08-10T16:16:31Z`
Route: `/data_control?referer=%2Ffeed%2Fpopular`
Version: `2021.07.27-5c76cdaa @ master`
<details>
<summary>Backtrace</summary>
<p>
   
```
Unexpected char 'K' at line 1, column 1 (JSON::ParseException)
  from /usr/share/crystal/src/json/lexer.cr:371:7 in 'raise'
  from /usr/share/crystal/src/json/lexer.cr:367:5 in 'unexpected_char'
  from /usr/share/crystal/src/json/lexer.cr:366:11 in 'unexpected_char'
  from /usr/share/crystal/src/slice.cr:229:5 in 'next_token'
  from /usr/share/crystal/src/json/parser.cr:13:12 in '->'
  from lib/kemal/src/kemal/route.cr:12:9 in '->'
  from /usr/share/crystal/src/primitives.cr:255:3 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from lib/kemal/src/kemal/filter_handler.cr:21:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/static_file_handler.cr:112:11 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from lib/kemal/src/kemal/init_handler.cr:12:7 in 'process'
  from /usr/share/crystal/src/http/server.cr:498:5 in '->'
  from /usr/share/crystal/src/primitives.cr:255:3 in 'run'
  from ???
```
</p>
</details>

[Also available on Snopyta PasteBin](https://bin.snopyta.org/?91bc88c04e266e77#3vdCruV713Xvm8KXYCuhjVZDvgMgGnDVkMj7iDdsdFd7)
Browser Link: https://yt.alt.s.dhusch.de/data_control?referer=%2Ffeed%2Fpopular

**Additional context**
I'm using YunoHost to host it

### PR DESCRIPTION
**Pull request by bbielsa**
Closes #2319 

1. Add a helper function `parse_subscription_export_csv()` to `helpers/utils.cr`, this function reads the body as CSV content, and parses the channel id column from the rows. 
2. Modify `invidious.cr` to perform a check on the body to determine if it is likely JSON (if so, the continue to use the current code for importing subscriptions). Otherwise, default to assuming it is in CSV format and parse the subscription export dump with the new helper function.

In the helper function, I used column index (in this case 0 corresponds to the `"Channel Id"` column), because I believe, based off of the bug report, that the header names are localized depending on the users language in this case "Kanal" = "Channel"

### CODE REVIEW

File: src/invidious.cr [hunk 1]
```diff
@@ -829,11 +829,14 @@ post "/data_control" do |env|
           user.subscriptions += subscriptions.xpath_nodes(%q(//outline[@type="rss"])).map do |channel|
             channel["xmlUrl"].match(/UC[a-zA-Z0-9_-]{22}/).not_nil![0]
           end
-        else
+        elsif body[0] == '['
```
[the remark this answers is not in the document]
Comment (bbielsa) [reply src/invidious.cr:832]: Would it be worth replacing these string comparison checks for determining content type and instead use the Content-Type header? https://crystal-lang.org/api/1.2.1/HTTP/FormData/Part.html#headers%3AHTTP%3A%3AHeaders-instance-method

[the remark this answers is not in the document]
Comment (SamantazFox) [reply src/invidious.cr:832]: mmmh, you're right!

[the remark this answers is not in the document]
Comment (bbielsa) [reply src/invidious.cr]: Fixed

File: src/invidious/helpers/utils.cr [hunk 2]
```diff
@@ -362,3 +363,23 @@ def fetch_random_instance
...
+
+    if channel_id.empty? 
+      next
+    end
+
+    subscriptions << channel_id
+  end
+
+  subscriptions
+end
```
[the remark this answers is not in the document]
Comment (bbielsa) [reply src/invidious/helpers/utils.cr]: Fixed

**Comment by unixfox** [comment]
In testing on https://yewtu.be

File: src/invidious/helpers/utils.cr [hunk 3]
```diff
@@ -1,4 +1,5 @@
 require "db"
+require "csv"
```
Comment (SamantazFox) [inline src/invidious/helpers/utils.cr:2]: this should be in `imports.cr`, with the code requiring it

File: src/invidious/user/imports.cr [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    # name, because the header name is localized depending on the
+    # language the user has set on their account
+    channel_id = row[0].strip
+
+    next if channel_id.empty?
+
+    subscriptions << channel_id
+  end
+
+  subscriptions
```
Comment (SamantazFox) [inline src/invidious/user/imports.cr:16]: please use a `return` statement here. While it's functionally the same thing, it makes the code more explicit :)

**Review by SamantazFox — CHANGES_REQUESTED** [review]

**Comment by unixfox** [comment]
Probably a related issue: https://github.com/iv-org/invidious/issues/2566
Need to be investigated.

**Comment by bbielsa** [comment]
I think it may be related to the logic for parsing the uploaded data. Currently the way it works is:

1. The POST handler checks the MIME type of the uploaded file
2. If the type matches `application/xml` or `text/xml`, then parse as XML
3. If the type matches `application/json`, then parse as JSON
4. Else, parse as CSV

I can explicitly check for the CSV MIME type, and I was going to do this originally, but it introduces the case where the user can upload any other type of file and there was no function to call that will return a simple user-readable error. What I can do is update the logic:

1. The POST handler checks the MIME type of the uploaded file
2. If the type matches `application/xml` or `text/xml`, then parse as XML
3. If the type matches `application/json`, then parse as JSON
4. If the type matches `text/csv`, then parse as CSV
5. Else, throw exception which shows the stack trace (may not be good for the user experience, but admittedly better than what I have created now)

A note for clarity: the "if type matches x" are the if statement cases in this commit https://github.com/iv-org/invidious/pull/2545/commits/0e138d64efe86cec4e9d70ddc0f49747fdc44a35

**Comment by unixfox** [comment]
Why not having multiples fields for csv and json?
Like this:
image

**Comment by SamantazFox** [comment]
> Why not having multiples fields for csv and json? Like this: image

This is confusing. Some users already complained that there are too many import options.

@bbielsa Is it possible to check the filename? so `if type == 'foo' || name.ends_with?(".foo")`.

**Comment by SamantazFox** [comment]
> Is it possible to check the filename? so `if type == 'foo' || name.ends_with?(".foo")`.

@bbielsa bump

**Comment by bbielsa** [comment]
> > Why not having multiples fields for csv and json? Like this:
> > image
> 
> This is confusing. Some users already complained that there are too many import options.
> 
> @bbielsa Is it possible to check the filename? so `if type == 'foo' || name.ends_with?(".foo")`.

I can check the file name, but in the event that the file is not XML, JSON, nor CSV what should I do? Is there an error page that I can redirect to? Or should I throw an exception and have the usual stack trace page show?

**Comment by bbielsa** [comment]
@SamantazFox Ok so minor problem, I am trying to return a 400 error when the user uploads an invalid file, but due to the way that the existing code processes the users uploads, its not actually possible to error out and return a 400 error at this point in the code, I believe.

Here is the relevant line of code (I have not committed it yet) https://gist.github.com/bbielsa/7d131aa2188945f591a8379ec0defc9b#file-invidious-cr-L19

In this case `next` goes to the next loop of the FormData loop `HTTP::FormData.parse(env.request) do |part|`, and then at the last line of the POST handler function it redirects to the referrer `env.redirect referer`, essentially overwriting the error I am trying to show the user. 

So I was able to add the changes you asked for and I explicitly check the file extension, but if the file uploaded is of an unrecognized type, it will not show the user and error and simply redirect to the last page, not the best user experience. 

Any ideas on what I can do to work around this? Some of the ideas I had resulted in some ugly code, so I'd appreciate any contributors input or opinions.

**Comment by SamantazFox** [comment]
> @SamantazFox Ok so minor problem, I am trying to return a 400 error when the user uploads an invalid file, but due to the way that the existing code processes the users uploads, its not actually possible to error out and return a 400 error at this point in the code, I believe.
> 
> Here is the relevant line of code (I have not committed it yet) https://gist.github.com/bbielsa/7d131aa2188945f591a8379ec0defc9b#file-invidious-cr-L19
> 
> In this case `next` goes to the next loop of the FormData loop `HTTP::FormData.parse(env.request) do |part|`, and then at the last line of the POST handler function it redirects to the referrer `env.redirect referer`, essentially overwriting the error I am trying to show the user.
> 
> So I was able to add the changes you asked for and I explicitly check the file extension, but if the file uploaded is of an unrecognized type, it will not show the user and error and simply redirect to the last page, not the best user experience.
> 
> Any ideas on what I can do to work around this? Some of the ideas I had resulted in some ugly code, so I'd appreciate any contributors input or opinions.

Hello ! Sorry for the late reply!

Kemal has a macro for returning a response early, even in blocks:
```
halt(env, status_code: 415,
  response: error_template(415, "Invalid subscription file uploaded")
)
```

Note that we're still using the `error_template` function to set the response's content (otherwise, there wouldn't be any HTML, just the error message specified).

(FYI: Code "415" is ["Unsupported Media Type (RFC 7231)"](https://en.wikipedia.org/wiki/HTTP_400#4xx_client_errors), which is more appropriate in this case ^^)

Also, can you please put any error message you add to the translations, with a key similar to this:
`error_message_invalid_subscription_file`? Thanks :)

**Comment by SamantazFox** [comment]
@bbielsa also note that currently (i.e with the latest commit in this PR and the code you provided as a gist) the CSV import doesn't work, the request times out. There is probably an issue with the way you use the CSV module...

**Comment by unixfox** [comment]
Removing from testing on https://yewtu.be

**Comment by SamantazFox** [comment]
@bbielsa bump! Do you plan to finish this PR soon?

**Comment by SamantazFox** [comment]
Deployed on https://test.invidious.io

**Comment by SamantazFox** [comment]
@matthewmcgarvey could you please review this ?

File: spec/invidious/user/imports_spec.cr [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+end
+
+Spectator.describe "Invidious::User::Imports" do
+  it "imports CSV" do
+    subscriptions = parse_subscription_export_csv(csv_sample)
+
+    expect(subscriptions).to be_an(Array(String))
+    expect(subscriptions.size).to eq(13)
+
+    expect(subscriptions).to contain("UC0hHW5Y08ggq-9kbrGgWj0A")
```
Comment (matthewmcgarvey) [inline spec/invidious/user/imports_spec.cr:35]: Just a note. `contain_exactly` is my preferred way of doing these kinds of asserts. It gives the benefit of saying it must have all these items without the drawback of having the match the order when trying to do `expect(subscriptions).to eq([...])`

https://gitlab.com/arctic-fox/spectator/-/wikis/Collection-Matchers#contain_exactly

**Review by matthewmcgarvey — APPROVED** [review]
Looks good 👍

**Review by matthewmcgarvey — APPROVED** [review]

**Pull request merged**